### PR TITLE
OOP Exercise/a-tiny-js-world updated

### PR DESCRIPTION
**Pull request by TaniaBondarenko**
OOP Exercise

[Demo](https://taniabondarenko.github.io/a-tiny-JS-world/)
[Code base](https://github.com/TaniaBondarenko/a-tiny-JS-world/blob/master/index.js)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/TaniaBondarenko/a-tiny-js-world-update/index.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  toString() {
+    return (` I\'m <b>${this.name}</b>, a ${this.gender}. To my friend ${this.friends} I always say: "${this.saying}"`);
+  }
+};
+
+class Person extends Inhabitant {
+  constructor(name, gender, saying, friends, species='human', legs=2, hands=2) {
+    super(name, gender, saying, friends)
+    this.species = species;
+    this.legs = legs;
```
Comment (OleksiyRudenko) [inline submissions/TaniaBondarenko/a-tiny-js-world-update/index.js:17]: Why `name`, `gender`, `saying` and `friends` are generalized into a base class, while `species` and `legs` (that are also properties shared by all inhabitants) are not?

File: submissions/TaniaBondarenko/a-tiny-js-world-update/index.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    return `Since I'm <em>${this.species}</em>,I have ${this.legs} legs and ${this.hands} hands.` + super.toString();
+  } 
+};
+
+class Animal extends Inhabitant {
+  constructor(name, gender, saying, species, friends, legs=4, hands=0) {
+    super(name, gender, saying, friends )
+    this.species = species;
+    this.legs = legs;
+    this.hands = hands;
```
Comment (OleksiyRudenko) [inline submissions/TaniaBondarenko/a-tiny-js-world-update/index.js:30]: Why do we need `hands` for animals that do not have hands by natural design?

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@TaniaBondarenko well done.
A few inconsistencies to fix yet.

File: submissions/TaniaBondarenko/a-tiny-js-world-update/index.js (hunk 1 above)
Comment (TaniaBondarenko) [reply submissions/TaniaBondarenko/a-tiny-js-world-update/index.js]: Thank you for. It's obvious)

File: submissions/TaniaBondarenko/a-tiny-js-world-update/index.js (hunk 2 above)
Comment (TaniaBondarenko) [reply submissions/TaniaBondarenko/a-tiny-js-world-update/index.js]: Done

**Comment by TaniaBondarenko** [comment]
Fixed and updated.

File: submissions/TaniaBondarenko/a-tiny-js-world-update/index.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
+class Inhabitant {
+  constructor(name, gender, saying, friends,legs, hands) {
+    this.name = name;
+    this.gender = gender;
+    this.saying = saying;
+    this.friends = friends;
+    this.legs = legs;
+    this.hands = hands;
```
Comment (OleksiyRudenko) [inline submissions/TaniaBondarenko/a-tiny-js-world-update/index.js:8]: Now `Animal` class inherits hands from `Inhabitant` class, thus all animals have hands.
Hands are unique to humans. So it is `Human` that has to introduce `hands` property and handle it solely.

File: submissions/TaniaBondarenko/a-tiny-js-world-update/index.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  }
+  toString() {
+    return (` I\'m <b>${this.name}</b>, a ${this.gender}. To my friend ${this.friends} I always say: "${this.saying}"`);
```
Comment (OleksiyRudenko) [inline submissions/TaniaBondarenko/a-tiny-js-world-update/index.js:11]: No good reasons for this class to let its children handle shared properties, not handling those by itself.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@TaniaBondarenko optimization went well, yet partially.

Every class should handle props it owns. That's how encapsulation principle is followed. 

There must be a good reason to allow child classes to break encapsulation principle by handling props owned by any of it ancestors.

File: submissions/TaniaBondarenko/a-tiny-js-world-update/index.js (hunk 3 above)
Comment (TaniaBondarenko) [reply submissions/TaniaBondarenko/a-tiny-js-world-update/index.js:8]: I think, I got it. Thank you.

File: submissions/TaniaBondarenko/a-tiny-js-world-update/index.js (hunk 4 above)
Comment (TaniaBondarenko) [reply submissions/TaniaBondarenko/a-tiny-js-world-update/index.js]: Understood. Fixed and updated.

File: submissions/TaniaBondarenko/a-tiny-js-world-update/index.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  }
+  toString() {
+    return (` ${this.legs} legs. I\'m <b>${this.name}</b>, a ${this.gender}. To my friend ${this.friends} I always say: "${this.saying}"`);
+  }
+};
+
+class Person extends Inhabitant {
+  constructor(name, gender, saying, friends, species='human', legs=2, hands=2) {
+    super(name, gender, saying, friends,legs)
+    this.species = species;
```
Comment (OleksiyRudenko) [inline submissions/TaniaBondarenko/a-tiny-js-world-update/index.js:17]: Remember one of the comments above?
This one:
> Why name, gender, saying and friends are generalized into a base class, while **species** and legs (that are also properties shared by all inhabitants) are not?

File: submissions/TaniaBondarenko/a-tiny-js-world-update/index.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  } 
+};
+
+class Creature extends Person{
+  constructor(name, gender, saying, friends, species = 'half-human', legs = 2, hands = 2) {
+    super (name, gender, saying, friends, species, legs=2, hands=2) 
+  }
+  toString() {
+    return super.toString();
+  }
```
Comment (OleksiyRudenko) [inline submissions/TaniaBondarenko/a-tiny-js-world-update/index.js:41]: A child class inhertis all properties and methods from its ancestors.
If we do not need to modify an inherited method behaviour then there is no need reassigning it.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@TaniaBondarenko some of the past code review comments are still valid though

File: submissions/TaniaBondarenko/a-tiny-js-world-update/index.js (hunk 6 above)
Comment (TaniaBondarenko) [reply submissions/TaniaBondarenko/a-tiny-js-world-update/index.js:41]: Got it! Thank you.

File: submissions/TaniaBondarenko/a-tiny-js-world-update/index.js (hunk 5 above)
Comment (TaniaBondarenko) [reply submissions/TaniaBondarenko/a-tiny-js-world-update/index.js]: I think, now I did it.

**Comment by TaniaBondarenko** [comment]
I suppose, that now the code looks as it should. Please check.
Thank you for your time.

**Review by OleksiyRudenko — APPROVED** [review]
@TaniaBondarenko well done!

**Comment by OleksiyRudenko** [comment]
@TaniaBondarenko did I miss a submission for https://github.com/kottans/frontend/blob/master/tasks/js-pre-oop.md ?

**Pull request merged**